framework/stats/stats.h

```diff
@@ -0,0 +1,172 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <ctime>
+#include <future>
+#include <map>
+#include <set>
+#include <vector>
+
+#include "common/error.h"
+
+#include "stats_common.h"
+#include "stats_provider.h"
+#include "timer.h"
+
+namespace vkb
+{
+class Device;
+class CommandBuffer;
+
+/*
+ * @brief Helper class for querying statistics about the CPU and the GPU
+ */
+class Stats
+{
+  public:
+	/**
+	 * @brief Constructs a Stats object
+	 * @param device Device on which to collect stats
+	 * @param num_framebuffers The number of buffers in the swapchain
+	 * @param requested_stats Set of stats to be collected if available
+	 * @param sampling_config Sampling mode configuration (polling or continuous)
+	 * @param buffer_size Size of the circular buffers
+	 */
+	Stats(Device &                   device,
+	      size_t                     num_framebuffers,
+	      const std::set<StatIndex> &requested_stats,
+	      CounterSamplingConfig      sampling_config = {CounterSamplingMode::Polling},
+	      size_t                     buffer_size     = 16);
+
+	/**
+	 * @brief Destroys the Stats object
+	 */
+	~Stats();
+
+	/**
+	 * @brief Resizes the stats buffers according to the width of the screen
+	 * @param width The width of the screen
+	 */
+	void resize(size_t width);
+
+	/**
+	 * @brief Checks if an enabled stat is available in the current platform
+	 * @param index The stat index
+	 * @return True if the stat is available, false otherwise
+	 */
+	bool is_available(StatIndex index) const;
+
+	/**
+	 * @brief Returns data relevant for graphing a specific statistic
+	 * @param index The stat index of the data requested
+	 * @return The data of the specified stat
+	 */
+	const StatGraphData &get_graph_data(StatIndex index) const;
+
+	/**
+	 * @brief Returns the collected data for a specific statistic
+	 * @param index The stat index of the data requested
+	 * @return The data of the specified stat
+	 */
+	const std::vector<float> &get_data(StatIndex index) const
+	{
+		return counters.at(index);
+	};
+
+	/**
+	 * @return The requested stats
+	 */
+	const std::set<StatIndex> &get_requested_stats() const
+	{
+		return requested_stats;
+	}
+
+	/**
+	 * @brief Update statistics, must be called after every frame
+	 */
+	void update(float delta_time, uint32_t active_frame_idx);
+
+	/**
+	 * @brief A command buffer that we want stats about has just begun
+	 * @param cb The command buffer
+ 	 * @param active_frame_idx Which of the framebuffers is active
+	 */
+	void command_buffer_begun(CommandBuffer &cb, uint32_t active_frame_idx);
```

Comment:
This function essentially prepares our providers to start collecting stats, I think this function name should be renamed to `begin` and the later `end`.

Reply:
The @brief should be changed respectively to detail this more accurately, along with when it should be called

Reply:
It's a tricky one, because for HWCPipe these are no-ops if I understood correctly (and they should be, we are reading system-wide counters), but for performance query we do need to collect data for a given command buffer

Reply:
Exactly. They don't begin or end anything in hwcpipe so calling the abstract interface begin and end doesn't feel right to me.